examples/connection_pools/src/main.rs

```diff
@@ -0,0 +1,198 @@
+#![feature(plugin, decl_macro)]
+#![plugin(rocket_codegen)]
+
+extern crate rocket;
+extern crate rocket_contrib;
+
+use std::io;
+use std::ops::Deref;
+use rocket::http::Status;
+use rocket::request::{self, FromRequest};
+use rocket::{Outcome, Request, State};
+use rocket::fairing::{Fairing, AdHoc};
+use rocket_contrib::databases::{self, DatabaseConfig, Poolable, diesel, postgres, r2d2, r2d2_postgres};
+
+// The two annotations would be part of the proc-macro crate to generate the
+// FromRequest implementation for the specified DB type.
+// #[derive(DbConn)]
+// #[database("my_sqlite_database")]
+struct TempStorage(r2d2::PooledConnection<diesel::r2d2::ConnectionManager<diesel::SqliteConnection>>);
+
+// #[derive(DbConn)]
+// #[database("my_primary_database")]
+struct PrimaryDb(r2d2::PooledConnection<r2d2_postgres::PostgresConnectionManager>);
+
+// #[derive(DbConn)]
+// #[database("my_warehouse_database")]
+struct WarehouseDb(r2d2::PooledConnection<r2d2_postgres::PostgresConnectionManager>);
+
+fn main() {
+    rocket::ignite()
+        .mount("/", routes![test_handler])
+        .attach(TempStorage::fairing())
+        .attach(PrimaryDb::fairing())
+        .attach(WarehouseDb::fairing())
+        .launch();
+}
+
+#[get("/")]
+fn test_handler(_storage: TempStorage, _primary: PrimaryDb, _warehouse: WarehouseDb) -> &'static str {
+    "Hello, Rocket with databases!"
+}
+
+// Code below here would be generated at compile-time via procedural macros
+struct TempStoragePool(r2d2::Pool<<diesel::SqliteConnection as Poolable>::Manager>);
```

Comment:
How are you going to go from reading `r2d2::PooledConnection<diesel::r2d2::ConnectionManager<diesel::SqliteConnection>>` inside the user's struct to generating `r2d2::Pool<<diesel::SqliteConnection as Poolable>::Manager>` inside the managed state struct? We can't assume that the internal type looks a certain way. For instance, if we were to use a type alias, inspecting the type wouldn't tell you much.

```rust
type PooledSqliteConn = r2d2::PooledConnection<diesel::r2d2::ConnectionManager<diesel::SqliteConnection>>;

#[database("my_sqlite_database")]
struct TempStorage(PooledSqliteConn);
```

The challenge is that we want all of this infrastructure to work for connections and databases we know nothing about, so we can't keep a static table of these things. Maybe the solution is to ask the user to write:

```rust
#[database("database_name")]
struct MyConnection(ConnectionType);

// Or, concretely:
#[database("my_sqlite_database")]
struct TempStorage(diesel::SqliteConnection);
```

Then we'll generate:

```rust
struct MyConnectionPool(r2d2::Pool<<ConnectionType as Poolable>::Manager>);

// Or, concretely:
struct TempStoragePool(r2d2::Pool<<diesel::SqliteConnection as Poolable>::Manager>);
```

The caveat is that user then can't do `.0` to access the internal connection. Maybe that's okay and this is where the `DbConn` trait comes in:

```rust
trait DbConn {
    type Connection = ConnectionType;
    fn inner(&self) -> &r2d2::PooledConnection<diesel::r2d2::ConnectionManager<Self::Connection>>;
}
```

The implementation for this would always be `&self.0`.

Are there cases when the stored connection type won't be `r2d2::PooledConnection<diesel::r2d2::ConnectionManager<Self::Connection>>`? If we use `r2d2`, I believe it will always be this, is that correct?

Reply:
Yes, that was becoming evident as I considered the procedural macro more and more that there was going to have to be a lookup table in the current form, which isn't ideal.

Unfortunately, non-Diesel Postgres connections don't follow that format. For example: `r2d2::PooledConnection<r2d2_postgres::PostgresConnectionManager>`. I need to check on the other connection pools we want to support, but I can verify how many variations of the `r2d2::PooledConnection<...>` type there will be.

Reply:
Given a connection type and a `Pooled` (or whatever we call it) trait implementation, we can retrieve the manager's type with:

```rust
<ConnectionType as Pooled>::Manager
```

Thus, we should be able to generate the following and have it work always:

```rust
// From:
#[database("database_name")]
struct MyConnection(ConnectionType);

// To:
struct MyConnection(r2d2::PooledConnection<<ConnectionType as Poolable>::Manager>);
struct MyConnectionPool(r2d2::Pool<<ConnectionType as Poolable>::Manager>);
```

Reply:
Yep, that would work. Also allows us to clean up the code the user writes to get the pool type generated.

👍